devito/function.py

```diff
@@ -816,61 +778,51 @@ class SparseFunction(CompositeFunction):
-                                        shape=(self.npoint, self.grid.dim),
-                                        space_order=0)
-            self._children.append(self.coordinates)
-            coordinates = kwargs.get('coordinates', None)
-            if coordinates is not None:
-                self.coordinates.data[:] = coordinates[:]
-
-            # A SparseFunction has no halo region
+            super(SparseFunction, self).__init__(*args, **kwargs)
+
+            npoint = kwargs.get('npoint')
+            if not isinstance(npoint, int) and npoint > 0:
+                raise ValueError('SparseFunction requires parameter `npoint` (> 0)')
+            self.npoint = npoint
+
+            # Grid must be provided
+            grid = kwargs.get('grid')
+            if kwargs.get('grid') is None:
+                raise ValueError('SparseFunction objects require a grid parameter')
+            self.grid = grid
+
+            self.dtype = kwargs.get('dtype', self.grid.dtype)
+            self.space_order = kwargs.get('space_order', 1)
+
+            # Set up coordinates of sparse points
+            coordinates = Function(name='%s_coords' % self.name,
+                                   dimensions=(self.indices[-1], Dimension(name='d')),
+                                   shape=(self.npoint, self.grid.dim), space_order=0)
+            coordinate_data = kwargs.get('coordinates')
+            if coordinate_data is not None:
+                coordinates.data[:] = coordinate_data[:]
+            self.coordinates = coordinates
+
+            # Halo region
             self._halo = tuple((0, 0) for i in range(self.ndim))
 
-    def __new__(cls, *args, **kwargs):
-        nt = kwargs.get('nt', 0)
-        npoint = kwargs.get('npoint')
-        kwargs['shape'] = (nt, npoint) if nt > 0 else (npoint, )
-
-        return Function.__new__(cls, *args, **kwargs)
+            # Padding region
+            self._padding = tuple((0, 0) for i in range(self.ndim))
 
     @classmethod
-    def _indices(cls, **kwargs):
-        """Return the default dimension indices for a given data shape
-
-        :return: indices used for axis.
-        """
-        dimensions = kwargs.get('dimensions', None)
-        grid = kwargs.get('grid', None)
-        nt = kwargs.get('nt', 0)
-        dim = Dimension(name='p')
-        indices = [grid.time_dim, dim] if nt > 0 else [dim]
-        return dimensions or indices
-
-    @property
-    def shape_domain(self):
+    def __indices_setup__(cls, **kwargs):
         """
-        Full allocated shape of the data associated with this
-        :class:`SparseFunction`.
+        Return the default dimension indices for a given data shape.
         """
-        return (self.nt, self.npoint) if self.nt > 0 else (self.npoint, )
+        dimensions = kwargs.get('dimensions')
+        if dimensions is not None:
+            return dimensions
+        else:
+            return (Dimension(name='p'),)
```

Comment:
Fixed it in my branch, but this makes the same default dimension for all sparse functions which is risky.

Reply:
flaw in the test suite then? I'll wait for your PR to fix it

Reply:
My PR is on top of yours, Ill fix it in following (only a problem when start taking FD that calls `new` without `dimension` I think)